Add ability to specify shared state and config directories

jeremyrickard opened the pull request:
**What this PR does / why we need it**:

This PR adds the ability to specify directories that should be created for %{_sharedstatedir} and  %{_sysconfdir}

This will ensure they are created during the install phase and that they show up in the files section of the resulting RPM spec.

In the spec, this would look like:

```
artifacts:
  createDirectories:
    config:
      kubernetes: 
        mode: 0755
      kubernetes/manifests: {}
    state:
      kubelet: {}
```

This would result in:

```
%install
mkdir -m 755 -p "%{buildroot}/%{_sysconfdir}/kubernetes"
mkdir -p "%{buildroot}/%{_sysconfdir}/kubernetes/manifests"
mkdir -p "%{buildroot}/%{_sharedstatedir}/kubelet"

%files
%dir %{_sysconfdir}/kubernetes
%dir %{_sysconfdir}/kubernetes/manifests
%dir %{_sharedstatedir}/kubelet
```


**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #239

**Special notes for your reviewer**:

adamperlin commented on docs/spec.schema.json:
```diff
@@ -57,6 +57,20 @@
+						"type": "string"
+					},
+					"type": "array",
+					"description": "ConfigDirectories is a list of directories the RPM should under %{_sysconfigdir}\t."
```
nit: "the RPM should place under"

jeremyrickard replied:
thank you!

adamperlin commented on docs/spec.schema.json:
```diff
@@ -57,6 +57,20 @@
+						"type": "string"
+					},
+					"type": "array",
+					"description": "SharedStateDirectories is a list of directories the RPM should under %{%_sharedstatedir}\t."
```
nit: "the RPM should place under"

jeremyrickard replied:
thank you!

adamperlin commented:
My only concern really is whether or not it's confusing to have `configDirectories` and `sharedStateDirectories` as artifacts, when they're things that are being created and not just extracted from the build. What do we think about this? Having the paths as lists instead of map keys does differentiate these sections of the `artifacts` from, say, the map keys.

jeremyrickard commented:
> My only concern really is whether or not it's confusing to have `configDirectories` and `sharedStateDirectories` as artifacts, when they're things that are being created and not just extracted from the build. What do we think about this? Having the paths as lists instead of map keys does differentiate these new sections of the `artifacts` from, say, the `binaries` section.

I think the `artifacts` section is really where we intend to put all the things that land in the artifact (i.e. the RPM) though, would it be more confusing to have extra pieces in different part of the spec file configuring / modifying that?

cpuguy83 commented on spec.go:
```diff
@@ -127,7 +127,12 @@ type Artifacts struct {
 	// Manpages is the list of manpages to include in the package.
 	Manpages map[string]ArtifactConfig `yaml:"manpages,omitempty" json:"manpages,omitempty"`
+	// ConfigDirectories is a list of directories the RPM should place under %{_sysconfigdir}	.
+	ConfigDirectories []string `yaml:"configDirectories,omitempty" json:"configDirectories,omitempty"`
```
Maybe we can do something like `createDirectories`

which is a type like so:

```go
type CreateArtifactDirectories struct {
    Config: map[string]ArtifactDirConfig // These go in /etc
    State: map[string]ArtifactDirConfig // These go in /var/lib
}
```

jeremyrickard replied:
@cpuguy83  what would you see modeled in map form?

```
artifacts:
     createDirectories: 
          config: 
              <what's the map look like here, how do we want to use subpath and name?>
         state:
               <what's the map look like here, how do we want to use subpath and name?>
    ```

cpuguy83 replied:
Map key would be the dir name to create.

jeremyrickard replied:
What would the other ArtifactConfig things be used for? or would we just pretty much never include them?  

```
createDirectories: 
          config: 
               kubernets: {}
               kubernetes/manifests: {}
```

or 

```
createDirectories: 
          config: 
               kubernets: 
                   subpath: manifests
```

jeremyrickard replied:
oh sorry, you said `ArtifactDirConfig` which could be other things my bad, sorry!

jeremyrickard replied:
maybe for artifact dir config we want to optionally specify mode (i.e. mkdir -m <specified>) ?

cpuguy83 replied:
Yep.

cpuguy83 commented on frontend/rpm/template.go:
```diff
@@ -326,6 +326,17 @@ func (w *specWrapper) Install() fmt.Stringer {
+	sort.Strings(w.Spec.Artifacts.ConfigDirectories)
+	for _, p := range w.Spec.Artifacts.ConfigDirectories {
+		dir := filepath.Join(`%{buildroot}/%{_sysconfdir}`, p)
+		fmt.Fprintln(b, "mkdir -p", dir)
```
Need to make sure the dirs are quoted for the shell.

```suggestion
		fmt.Fprintf(b, "mkdir -p %q\n", dir)
```

jeremyrickard replied:
good catch thanks!

cpuguy83 commented on frontend/rpm/template.go:
```diff
@@ -326,6 +326,17 @@ func (w *specWrapper) Install() fmt.Stringer {
+	sort.Strings(w.Spec.Artifacts.SharedStateDirectories)
+	for _, p := range w.Spec.Artifacts.SharedStateDirectories {
+		dir := filepath.Join(`%{buildroot}/%{_sharedstatedir}`, p)
+		fmt.Fprintln(b, "mkdir -p", dir)
```
Same, dir needs quoting.

adamperlin commented:
> > My only concern really is whether or not it's confusing to have `configDirectories` and `sharedStateDirectories` as artifacts, when they're things that are being created and not just extracted from the build. What do we think about this? Having the paths as lists instead of map keys does differentiate these new sections of the `artifacts` from, say, the `binaries` section.
> 
> I think the `artifacts` section is really where we intend to put all the things that land in the artifact (i.e. the RPM) though, would it be more confusing to have extra pieces in different part of the spec file configuring / modifying that?

Yeah, that makes sense. It would likely be much more confusing to have extra pieces in other parts of the spec. The `createDirectories` idea seems good since it makes the creation piece explicit

jeremyrickard commented:
OK, I have refactored this to use the suggestion above to have a `createDirectories` block in the YAML and corresponding struct in the spec.go. 

Now the user's YAML spec  file would look like:

```
artifacts:
  binaries:
    kubernetes/_output/bin/kubelet: {}
  createDirectories:
    config:
      kubernetes: 
        mode: 0755
      kubernetes/manifests: {}
    state:
      kubelet: {}
```

if `mode` isn't provided, the `-m` arg will be committed from the mkdir. Path is now also quoted.

cpuguy83 commented on spec.go:
```diff
@@ -127,9 +127,27 @@ type Artifacts struct {
+// CreateArtifactDirectories describes various directories that should be created on install.
+// CreateArtifactDirectories represents different directory paths that are common to RPM systems.
+type CreateArtifactDirectories struct {
+	// Config is a list of directories the RPM should place under %{_sysconfigdir} (i.e. /etc)
```
```suggestion
	// Config is a list of directories the RPM should place under the system config directory (e.g. `/etc`)
```

cpuguy83 commented on spec.go:
```diff
@@ -127,9 +127,27 @@ type Artifacts struct {
+type CreateArtifactDirectories struct {
+	// Config is a list of directories the RPM should place under %{_sysconfigdir} (i.e. /etc)
+	Config map[string]ArtifactDirConfig
+	// State is a list of directories the RPM should place under %{%_sharedstatedir} (i.e. /var/lib).
```
```suggestion
	// State is a list of directories the RPM should place under the common directory for shared (e.g. `/var/lib`).
```

cpuguy83 reviewed (changes requested):
Can you add a test case in https://github.com/Azure/dalec/blob/565c033351b7d5d104436fdafe56e887ecc00ae8/test/mariner2_test.go#L21?

jeremyrickard commented:
@cpuguy83  I added a test. I noticed that we removed the previous tests that checked permissions, in this case I think we should probably check them, although I am not sure if this is portable (paths too maybe?).

I did a first version of this that just ensured it ran successfully and that was fine but I thought maybe we would want to test the permissions because it's an option that can be specified in the config for the directories. LMK what you think.

cpuguy83 commented:
>  in this case I think we should probably check them, although I am not sure if this is portable (paths too maybe?)

We'll see when I implement this in jammy 😅
